File: src/CSharp/StoryGenerator.Tests/Examples/UserProfile.cs
namespace StoryGenerator.Tests.Examples;

/// <summary>
/// Immutable user profile record.
/// Demonstrates C# best practices: using records for DTOs/value objects.
/// </summary>
public record UserProfile
{
    /// <summary>
    /// Gets the username.
    /// </summary>
    public required string Username { get; init; }

    /// <summary>
    /// Gets the email address.
    /// </summary>
    public required string Email { get; init; }

    /// <summary>
    /// Gets the full name.
    /// </summary>
    public required string FullName { get; init; }

    /// <summary>
    /// Gets the creation timestamp.
    /// </summary>
    public required DateTime CreatedAt { get; init; }

    /// <summary>
    /// Gets the optional biography.
    /// </summary>
    public string? Bio { get; init; }

    /// <summary>
    /// Initializes a new instance of the <see cref="UserProfile"/> record.
    /// Private constructor to enforce use of factory method.
    /// </summary>
    private UserProfile()
    {
    }

    /// <summary>
    /// Factory method to create a validated UserProfile.
    /// </summary>
    /// <param name="username">Unique username for the user.</param>
    /// <param name="email">User's email address.</param>
    /// <param name="fullName">User's full name.</param>
    /// <param name="createdAt">Timestamp of profile creation.</param>
    /// <param name="bio">Optional biography text.</param>
    /// <returns>A validated UserProfile instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if any required parameter is null.</exception>
    /// <exception cref="ArgumentException">Thrown if validation fails.</exception>
    public static UserProfile Create(
        string username,
        string email,
        string fullName,
        DateTime createdAt,
        string? bio = null)
    {
        ArgumentNullException.ThrowIfNull(username);
        ArgumentNullException.ThrowIfNull(email);
        ArgumentNullException.ThrowIfNull(fullName);

        if (username.Length < 3)
        {
            throw new ArgumentException("Username must be at least 3 characters", nameof(username));
        }

        if (!IsValidEmail(email))
        {
            throw new ArgumentException("Invalid email address", nameof(email));
        }

        if (string.IsNullOrWhiteSpace(fullName))
        {
            throw new ArgumentException("Full name is required", nameof(fullName));
        }

        return new UserProfile
        {
            Username = username,
            Email = email,
            FullName = fullName,
            CreatedAt = createdAt,
            Bio = bio
        };
    }

    /// <summary>
    /// Gets the display name for the user.
    /// </summary>
    /// <returns>Full name followed by username in parentheses.</returns>
    public string GetDisplayName() => $"{FullName} (@{Username})";

    /// <summary>
    /// Checks if the user has a biography.
    /// </summary>
    /// <returns>True if bio is set and not empty, false otherwise.</returns>
    public bool HasBio() => !string.IsNullOrWhiteSpace(Bio);

    /// <summary>
    /// Validates an email address using MailAddress constructor.
    /// </summary>
    /// <param name="email">The email address to validate.</param>
    /// <returns>True if the email is valid, false otherwise.</returns>
    private static bool IsValidEmail(string email)
    {
        if (string.IsNullOrWhiteSpace(email))
        {
            return false;
        }

        try
        {
            var mailAddress = new System.Net.Mail.MailAddress(email);
            return mailAddress.Address == email;
        }
        catch
        {
            return false;
        }
    }
}

File: src/Python/models/user_profile.py
"""User profile models demonstrating immutable dataclasses.

TDD example showcasing Python best practices:
- Frozen dataclasses for immutability
- Type hints
- Validation in __post_init__
- Clear documentation
"""

from dataclasses import dataclass
from datetime import datetime
from typing import Optional


@dataclass(frozen=True)
class UserProfile:
    """Immutable user profile with validation.

    This demonstrates Step 3 of Python best practices: using frozen dataclasses
    for value objects and DTOs to ensure immutability.

    Attributes:
        username: Unique username for the user.
        email: User's email address.
        full_name: User's full name.
        created_at: Timestamp of profile creation.
        bio: Optional biography text.
    """

    username: str
    email: str
    full_name: str
    created_at: datetime
    bio: Optional[str] = None

    def __post_init__(self) -> None:
        """Validate profile data after initialization.

        Raises:
            ValueError: If any validation fails.
        """
        if not self.username or len(self.username) < 3:
            raise ValueError("Username must be at least 3 characters")

        if not self.email or "@" not in self.email:
            raise ValueError("Invalid email address")

        if not self.full_name:
            raise ValueError("Full name is required")

    def get_display_name(self) -> str:
        """Get the display name for the user.

        Returns:
            Full name followed by username in parentheses.

        Example:
            >>> profile = UserProfile(
            ...     username="johndoe",
            ...     email="john@example.com",
            ...     full_name="John Doe",
            ...     created_at=datetime.now()
            ... )
            >>> profile.get_display_name()
            'John Doe (@johndoe)'
        """
        return f"{self.full_name} (@{self.username})"

    def has_bio(self) -> bool:
        """Check if user has a biography.

        Returns:
            True if bio is set and not empty, False otherwise.
        """
        return self.bio is not None and len(self.bio.strip()) > 0
